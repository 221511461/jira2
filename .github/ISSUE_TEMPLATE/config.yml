blank_issues_enabled: false
contact_links:
  - name: Support and Bug Requests
    url: https://support.atlassian.com/jira-software-cloud/
    about: Need some help? Please submit a support request or bug report via Jira Software Support portal
