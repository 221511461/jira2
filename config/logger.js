const bunyan = require('bunyan');
const bformat = require('bunyan-format');
// add levelInString to include DEBUG | ERROR | INFO | WARN
const formatOut = bformat({ outputMode: 'bunyan', levelInString: true });

const logger = bunyan.createLogger(
  {
    name: 'github-for-jira',
    stream: formatOut,
    level: 'info',
    serializers: {
      req: require('bunyan-express-serializer'),
    },
  },
);

// Suppress logging in tests
if (process.env.NODE_ENV === 'test') {
  logger.level(bunyan.FATAL + 1);
}

console.debug = logger.debug.bind(logger);
console.error = logger.error.bind(logger);
console.log = logger.info.bind(logger);
console.warn = logger.warn.bind(logger);

module.exports = logger;
