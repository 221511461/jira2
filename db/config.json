{
  "development": {
    "database": "jira-dev",
    "host": "127.0.0.1",
    "dialect": "postgres"
  },
  "test": {
    "database": "jira-test",
    "host": "127.0.0.1",
    "dialect": "postgres"
  },
  "production": {
    "dialect": "postgres",
    "use_env_variable": "PG_DATABASE_URL",
    "disable_sql_logging": true,
    "ssl": false,
    "dialectOptions": {
      "ssl": false
    },
    "pool": {
      "max": 15
    }
  }
}
