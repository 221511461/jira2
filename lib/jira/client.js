const axios = require('axios')
const getJiraDetails = require('./secret')
const jwt = require('atlassian-jwt')
const querystring = require('querystring')

/*
 * The Atlassian API uses JSON Web Tokens (JWT) for authentication along with
 * Query String Hashing (QSH) to prevent URL tampering. IAT, or issued-at-time,
 * is a Unix-style timestamp of when the token was issued. EXP, or expiration
 * time, is a Unix-style timestamp of when the token expires and must be no
 * more than three minutes after the IAT. Since our tokens are per-request and
 * short-lived, we use a timeout of 30 seconds.
 */
function getExpirationInSeconds () {
  const nowInSeconds = Math.floor(Date.now() / 1000)

  return {
    iat: nowInSeconds,
    exp: nowInSeconds + 30
  }
}

module.exports = (context) => {
  const { secret, baseURL } = getJiraDetails()

  function request (method, path, query = {}, data = undefined, options = {}) {
    const url = baseURL + '/rest/api/latest' + path

    const jwtToken = jwt.encode({
      ...getExpirationInSeconds(),
      iss: 'com.github.integration',
      qsh: jwt.createQueryStringHash({
        method,
        originalUrl: url,
        query
      })
    }, secret)

    return axios({
      url: `${url}?${querystring.stringify(query)}`,
      baseURL,
      method,
      data,
      headers: {
        Authorization: `JWT ${jwtToken}`,
        ...options.headers
      }
    })
  }

  return {
    baseURL,
    issues: {
      get: (issueId, query) => request('GET', `/issue/${issueId}`, query)
    }
  }
}
