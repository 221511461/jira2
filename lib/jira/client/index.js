const { Installation } = require('../../models')
const getAxiosInstance = require('./axios')
const { getJiraId } = require('../util/id')

// Max number of issue keys we can pass to the Jira API
const ISSUE_KEY_API_LIMIT = 100

/**
 * Deduplicates elements in an array
 */
const dedup = (array) => [...new Set(array)]

/**
 * Deduplicates the issueKeys field for branches and commits
 */
const dedupIssueKeys = (resources) => {
  resources.forEach(resource => {
    resource.issueKeys = dedup(resource.issueKeys)
  })
  return resources
}

/**
 * Returns if the max length of the issue
 * key field is within the limit
 */
const withinIssueKeyLimit = (resources) => {
  if (resources == null) return []

  const issueKeyCounts = resources.map((resource) => resource.issueKeys.length)
  return Math.max(...issueKeyCounts) <= ISSUE_KEY_API_LIMIT
}

/*
 * Similar to the existing Octokit rest.js instance included in probot
 * apps by default, this client adds a Jira client that allows us to
 * abstract away the underlying HTTP requests made for each action. In
 * general, the client should match the Octokit rest.js design for clear
 * interoperability.
 */
module.exports = async (id, installationId, jiraHost) => {
  const { jiraHost: baseURL, sharedSecret: secret } = await Installation.getForHost(jiraHost)

  const instance = getAxiosInstance(id, installationId, baseURL, secret)

  const client = {
    baseURL: instance.defaults.baseURL,
    issues: {
      // eslint-disable-next-line camelcase
      get: (issue_id, query = { fields: 'summary' }) => instance.get(`/rest/api/latest/issue/:issue_id`, {
        fields: {
          ...query,
          issue_id
        }
      }),
      getAll: async (issueIds, query) => (await Promise.all(issueIds.map(issueId => client.issues.get(issueId, query).catch(error => error))))
        .filter(response => response.status === 200)
        .map(response => response.data),
      parse: (text) => {
        const jiraIssueRegex = /[A-Z]+-[0-9]+/g
        if (!text) return null
        return text.match(jiraIssueRegex)
      },
      comments: {
        // eslint-disable-next-line camelcase
        getForIssue: (issue_id) => instance.get(`/rest/api/latest/issue/:issue_id/comment`, {
          fields: {
            issue_id
          }
        }),
        // eslint-disable-next-line camelcase
        addForIssue: (issue_id, payload) => instance.post(`/rest/api/latest/issue/:issue_id/comment`, payload, {
          fields: {
            issue_id
          }
        })
      },
      transitions: {
        // eslint-disable-next-line camelcase
        getForIssue: (issue_id) => instance.get(`/rest/api/latest/issue/:issue_id/transitions`, {
          fields: {
            issue_id
          }
        }),
        // eslint-disable-next-line camelcase
        updateForIssue: (issue_id, transition_id) => instance.post(`/rest/api/latest/issue/:issue_id/transitions`, {
          transition: {
            id: transition_id
          }
        }, {
          fields: {
            issue_id
          }
        })
      },
      worklogs: {
        // eslint-disable-next-line camelcase
        getForIssue: (issue_id) => instance.get(`/rest/api/latest/issue/:issue_id/worklog`, {
          fields: {
            issue_id
          }
        }),
        // eslint-disable-next-line camelcase
        addForIssue: (issue_id, payload) => instance.post(`/rest/api/latest/issue/:issue_id/worklog`, payload, {
          fields: {
            issue_id
          }
        })
      }
    },
    devinfo: {
      branch: {
        delete: (repositoryId, branchRef) => instance.delete(`/rest/devinfo/0.10/repository/${repositoryId}/branch/${getJiraId(branchRef)}`, {
          fields: { _updateSequenceId: Date.now() }
        })
      },
      // Add methods for handling installationId properties that exist in Jira
      installation: {
        exists: (gitHubInstallationId) => instance.get(`/rest/devinfo/0.10/existsByProperties?installationId=${gitHubInstallationId}`),
        delete: (gitHubInstallationId) => instance.delete(`/rest/devinfo/0.10/bulkByProperties?installationId=${gitHubInstallationId}`)
      },
      // Migration endpoints do not take any parameters,
      // but return 500 errors if the body is empty or null.
      // Passing an empty object gets around this issue.
      migration: {
        complete: () => instance.post('/rest/devinfo/0.10/github/migrationComplete', {}),
        undo: () => instance.post('/rest/devinfo/0.10/github/undoMigration', {})
      },
      pullRequest: {
        delete: (repositoryId, number) => instance.delete(`/rest/devinfo/0.10/repository/${repositoryId}/pull_request/${number}`, {
          fields: { _updateSequenceId: Date.now() }
        })
      },
      repository: {
        get: (repositoryId) => instance.get(`/rest/devinfo/0.10/repository/${repositoryId}`),
        delete: (repositoryId) => instance.delete(`/rest/devinfo/0.10/repository/${repositoryId}`, {
          fields: { _updateSequenceId: Date.now() }
        }),
        update: (data, options) => {
          // Deduplicate issue keys in commits and branches before request
          if ('commits' in data) data.commits = dedupIssueKeys(data.commits)
          if ('branches' in data) {
            data.branches = dedupIssueKeys(data.branches)
            data.branches.forEach(branch => {
              if ('lastCommit' in branch) {
                branch.lastCommit = dedupIssueKeys([branch.lastCommit])[0]
              }
            })
          }

          if (withinIssueKeyLimit(data.commits) && withinIssueKeyLimit(data.branches)) {
            instance.post('/rest/devinfo/0.10/bulk', {
              preventTransitions: (options && options.preventTransitions) || false,
              repositories: [data],
              properties: {
                installationId
              }
            })
          } else {
            // We have more than 100 issue keys so raise an exception
            throw new Error('A commit exceeded referenced issue key limit')
          }
        }
      }
    }
  }

  return client
}
