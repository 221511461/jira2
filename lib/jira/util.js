const jiraIssueReferenceRegex = /\[([A-Z]+-[0-9]+)\]/g
const jiraIssueRegex = /[A-Z]+-[0-9]+/g

module.exports = function (context) {
  async function findJiraIssueReferences (text, shouldValidate = true) {
    const issues = text.match(jiraIssueRegex)

    if (!shouldValidate) {
      return issues
    }

    const fetchedIssues = await Promise.all(
      issues.map(issue =>
        context.jira.issues.get(issue, {
          fields: 'summary'
        }).catch(error => error)
      )
    )

    return fetchedIssues
      .filter(response => response.status === 200)
      .map(response => response.data.key)
  }

  async function addJiraIssueLinks (text, shouldValidate = true) {
    const issues = await findJiraIssueReferences(text, shouldValidate)

    return text.replace(jiraIssueReferenceRegex, (match, issueKey) => {
      if (shouldValidate && !issues.includes(issueKey)) {
        return match
      }

      return `[${issueKey}](${context.jira.baseURL}/browse/${issueKey})`
    })
  }

  return {
    addJiraIssueLinks
  }
}
