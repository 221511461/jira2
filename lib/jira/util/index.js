const parseJiraIssueReferences = require('./parse-jira-issue-references')

const jiraIssueReferenceRegex = /\[([A-Z]+-[0-9]+)\](?!\()/g

module.exports = function (context) {
  async function fetchJiraIssues (issueReferences) {
    const fetchedIssues = await Promise.all(
      issueReferences.map(issueReference =>
        context.jira.issues.get(issueReference, {
          fields: 'summary'
        }).catch(error => error)
      )
    )

    return fetchedIssues
      .filter(response => response.status === 200)
      .map(response => response.data)
  }

  async function addJiraIssueLinks (text) {
    const issueReferences = parseJiraIssueReferences(text)
    const issues = (await fetchJiraIssues(issueReferences))
      .reduce((issues, issue) => ({
        ...issues,
        [issue.key]: issue
      }), {})

    return text.replace(jiraIssueReferenceRegex, (match, issueKey) => {
      if (!issues[issueKey]) {
        return match
      }

      return `[${issueKey} ${issues[issueKey].fields.summary}](${context.jira.baseURL}/browse/${issueKey})`
    })
  }

  return {
    addJiraIssueLinks
  }
}
