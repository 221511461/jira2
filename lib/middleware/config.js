const getConfig = require('probot-config')

module.exports = async (context) => {
  context.config = await getConfig(context, 'jira.yml')

  return context
}
