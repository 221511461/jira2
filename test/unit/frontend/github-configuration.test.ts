import supertest from "supertest";
import { Installation, Subscription } from "../../../src/models";
import SubscriptionClass from "../../../src/models/subscription";
import FrontendApp from "../../../src/frontend/app";
import { getLogger } from "../../../src/config/logger";
import express, { Application } from "express";
import { getSignedCookieHeader } from "../util/cookies";
import { ViewerRepositoryCountQuery } from "../../../src/github/client/github-queries";

jest.mock("../../../src/config/feature-flags");

describe("Github Configuration", () => {
	let frontendApp: Application;
	let sub: SubscriptionClass;

	const authenticatedUserResponse = { login: "test-user" };
	const adminUserResponse = { login: "admin-user" };
	const organizationMembershipResponse = { role: "member" };
	const organizationAdminResponse = { role: "admin" };
	const userInstallationsResponse = {
		total_count: 2,
		installations: [
			{
				account: {
					login: "test-org"
				},
				id: 1,
				target_type: "Organization"
			},
			{
				id: 3
			}
		]
	};

	beforeEach(async () => {
		sub = await Subscription.create({
			gitHubInstallationId: 15,
			jiraHost,
			jiraClientKey: "myClientKey"
		});

		await Installation.create({
			jiraHost,
			clientKey: "abc123",
			secrets: "def234",
			sharedSecret: "ghi345"
		});

		frontendApp = express();
		frontendApp.use((request, _, next) => {
			request.log = getLogger("test");
			next();
		});
		frontendApp.use(FrontendApp({
			getSignedJsonWebToken: () => "token",
			getInstallationAccessToken: async () => "access-token"
		}));
	});

	afterEach(async () => {
		await Installation.destroy({ truncate: true });
		await Subscription.destroy({ truncate: true });
	});

	describe("Github Token Validation", () => {
		it("should return redirect to github oauth flow for GET request if token is missing", async () =>
			supertest(frontendApp)
				.get("/github/configuration")
				.set(
					"Cookie",
					getSignedCookieHeader({
						jiraHost
					})
				)
				.expect(res => {
					expect(res.status).toBe(302);
					expect(res.headers.location).toContain("github.com/login/oauth/authorize");
				}));

		it("should return redirect to github oauth flow for GET request if token is invalid", async () => {
			githubNock
				.get("/")
				.matchHeader("Authorization", /^Bearer .+$/)
				.reply(403);

			return supertest(frontendApp)
				.get("/github/configuration")
				.set(
					"Cookie",
					getSignedCookieHeader({
						jiraHost,
						githubToken: "token"
					})
				)
				.expect(res => {
					expect(res.status).toBe(302);
					expect(res.headers.location).toContain("github.com/login/oauth/authorize");
				});
		});

		it("should return 401 if doing a POST request with a missing github token", async () =>
			supertest(frontendApp)
				.post("/github/configuration")
				.set(
					"Cookie",
					getSignedCookieHeader({
						jiraHost
					})
				)
				.expect(401));

		it("should return 401 if doing a POST request with an invalid github token", async () => {
			githubNock
				.get("/")
				.matchHeader("Authorization", /^Bearer .+$/)
				.reply(403);

			await supertest(frontendApp)
				.post("/github/configuration")
				.set(
					"Cookie",
					getSignedCookieHeader({
						jiraHost,
						githubToken: "token"
					})
				)
				.expect(401);
		});
	});

	describe("#GET", () => {
		it("should return 200 when calling with valid Github Token", async () => {
			githubNock
				.get("/")
				.matchHeader("Authorization", /^(Bearer|token) .+$/i)
				.reply(200);

			githubNock
				.get("/user")
				.reply(200, { login: "test-user" });

			githubNock
				.post(`/app/installations/${sub.gitHubInstallationId}/access_tokens`)
				.reply(200, {
					token: "token",
					expires_at: Date.now() + 999999
				});

			githubNock
				.get(`/app/installations/${sub.gitHubInstallationId}`)
				.twice()
				.reply(200, {
					"id": 1,
					"account": {
						"login": "octocat",
						"id": 1,
						"type": "User"
					},
					"html_url": "https://github.com/organizations/github/settings/installations/1",
					"target_type": "Organization",
					"created_at": "2017-07-08T16:18:44-04:00",
					"updated_at": "2017-07-08T16:18:44-04:00"
				});

			githubNock
				.get(`/user/installations`)
				.reply(200, {
					installations: [{
						id: sub.gitHubInstallationId,
						account: {
							login: "test-org"
						},
						target_type: "Organization"
					}]
				});

			githubNock
				.get(`/orgs/test-org/memberships/test-user`)
				.reply(200, {
					role: "admin"
				});

			githubNock
				.get(`/app`)
				.reply(200, {
					html_url: "https://github.com/apps/jira"
				});

			githubNock
				.post("/graphql", { query: ViewerRepositoryCountQuery })
				.query(true)
				.reply(200, {
					data: {
						viewer: {
							repositories: {
								totalCount: 1
							}
						}
					}
				});


			await supertest(frontendApp)
				.get("/github/configuration")
				.set(
					"Cookie",
					getSignedCookieHeader({
						jiraHost,
						githubToken: "token"
					})
				)
				.expect(200);
		});

		it("should return 200 even when organization is IP blocked", async () => {
			githubNock
				.get("/")
				.matchHeader("Authorization", /^(Bearer|token) .+$/i)
				.reply(200);

			githubNock
				.get("/user")
				.reply(200, { login: "test-user" });

			githubNock
				.post(`/app/installations/${sub.gitHubInstallationId}/access_tokens`)
				.reply(200, {
					token: "token",
					expires_at: Date.now() + 999999
				});

			githubNock
				.get(`/app/installations/${sub.gitHubInstallationId}`)
				.twice()
				.reply(403, {
					message: "Although you appear to have the correct authorization credentials, the `Fusion-Arc` organization has an IP allow list enabled, and 13.52.4.51 is not permitted to access this resource."
				});

			githubNock
				.get("/user/installations")
				.reply(200, {
					installations: [{
						id: sub.gitHubInstallationId,
						account: {
							login: "test-org"
						},
						target_type: "Organization"
					}]
				});

			githubNock
				.get("/orgs/test-org/memberships/test-user")
				.reply(200, {
					role: "admin"
				});

			githubNock
				.get("/app")
				.reply(200, {
					html_url: "https://github.com/apps/jira"
				});

			githubNock
				.post("/graphql", { query: ViewerRepositoryCountQuery })
				.query(true)
				.reply(403,  {
					message: "Although you appear to have the correct authorization credentials, the `Fusion-Arc` organization has an IP allow list enabled, and 13.52.4.51 is not permitted to access this resource."
				});

			await supertest(frontendApp)
				.get("/github/configuration")
				.set(
					"Cookie",
					getSignedCookieHeader({
						jiraHost,
						githubToken: "token"
					})
				)
				.expect(200);
		});
	});

	describe("#POST", () => {
		it("should return a 401 if no GitHub token present in session", () =>
			supertest(frontendApp)
				.post("/github/configuration")
				.send({})
				.set(
					"Cookie",
					getSignedCookieHeader({ jiraHost })
				)
				.expect(401));

		it("should return a 401 if no Jira host present in session", () =>
			supertest(frontendApp)
				.post("/github/configuration")
				.send({})
				.set(
					"Cookie",
					getSignedCookieHeader({
						githubToken: "test-github-token"
					})
				)
				.expect(401));

		it("should return a 401 if the user doesn't have access to the requested installation ID", async () => {
			// This is for github token validation check
			githubNock
				.get("/")
				.matchHeader("Authorization", /^(Bearer|token) .+$/i)
				.reply(200);

			githubNock
				.get("/user/installations")
				.reply(200, userInstallationsResponse);

			await supertest(frontendApp)
				.post("/github/configuration")
				.send({
					installationId: 2
				})
				.type("form")
				.set(
					"Cookie",
					getSignedCookieHeader({
						githubToken: "test-github-token",
						jiraHost
					})
				)
				.expect(401);
		});

		it("should return a 401 if the user is not an admin of the Org", () => {
			// This is for github token validation check
			githubNock
				.get("/")
				.matchHeader("Authorization", /^(Bearer|token) .+$/i)
				.reply(200);

			githubNock
				.get("/user/installations")
				.reply(200, userInstallationsResponse);
			githubNock
				.get("/user")
				.reply(200, authenticatedUserResponse);
			githubNock
				.get("/orgs/test-org/memberships/test-user")
				.reply(200, organizationMembershipResponse);
			return supertest(frontendApp)
				.post("/github/configuration")
				.send({
					installationId: 1
				})
				.type("form")
				.set(
					"Cookie",
					getSignedCookieHeader({
						githubToken: "test-github-token",
						jiraHost
					})
				)
				.expect(401);
		});

		it("should return a 400 if no installationId is present in the body", () => {
			// This is for github token validation check
			githubNock
				.get("/")
				.matchHeader("Authorization", /^(Bearer|token) .+$/i)
				.reply(200);
			return supertest(frontendApp)
				.post("/github/configuration")
				.send({})
				.set(
					"Cookie",
					getSignedCookieHeader({
						githubToken: "test-github-token",
						jiraHost
					})
				)
				.expect(400);
		});

		it("should return a 200 and install a Subscription", async () => {
			const jiraHost = "test-jira-host";

			// This is for github token validation check
			githubNock
				.get("/")
				.matchHeader("Authorization", /^(Bearer|token) .+$/i)
				.reply(200);

			githubNock
				.get("/user/installations")
				.reply(200, userInstallationsResponse);
			githubNock
				.get("/user")
				.reply(200, adminUserResponse);
			githubNock
				.get("/orgs/test-org/memberships/admin-user")
				.reply(200, organizationAdminResponse);

			const jiraClientKey = "a-unique-client-key";

			await supertest(frontendApp)
				.post("/github/configuration")
				.send({
					installationId: 1,
					clientKey: jiraClientKey
				})
				.type("form")
				.set(
					"Cookie",
					getSignedCookieHeader({
						githubToken: "test-github-token",
						jiraHost
					})
				)
				.expect(200);
		});
	});
});
