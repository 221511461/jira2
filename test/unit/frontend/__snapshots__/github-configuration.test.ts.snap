// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frontend GitHub Configuration #post should return a 200 and install a Subscription 1`] = `"jira.v0.Action"`;

exports[`Frontend GitHub Configuration #post should return a 200 and install a Subscription 2`] = `
Object {
  "actionSource": "web_console",
  "association": "subscription",
  "githubInstallationId": 1,
  "jiraHostname": "test-jira-host",
  "type": "created",
}
`;
